### ISSUE TITLE
Bottom border is cut off when pager is active on Windows

### ISSUE THREAD

**Issue by brainplot** [sharkdp/bat#1098]
**What version of `bat` are you using?**
`bat 0.15.4`

**Describe the bug you encountered:**
It looks like the bottom border `bat` displays by default is cut off on Windows, when the pager is active, as shown in this screenshot:

#### _With pager_
![image](https://user-images.githubusercontent.com/28014843/86616902-c9dedb00-bfb6-11ea-843a-10374dea75e1.png)

#### _Without pager_
![image](https://user-images.githubusercontent.com/28014843/86617326-59848980-bfb7-11ea-993d-09df584b008b.png)

I have `bat` also installed in WSL on Ubuntu 20.04 and this issue isn't present at all, which makes me think there's something wrong with how Windows handles `less`.

**Describe what you expected to happen?**
Bottom border not being cut off like on other platforms :)

**How did you install `bat`?**
`scoop install bat`

**EDIT**: I'll add my `less` version for completeness
```
less 561 (Spencer V8 regular expressions)
Copyright (C) 1984-2020  Mark Nudelman
```

---

Windows version; Windows 10 2004 (OS Build 19041.329)

**Comment by sharkdp** [comment]
This does look like a problem with the `--quit-if-one-screen` option in your version of `less`.

As a workaround, you could try to use `--pager="less -R"` in your `bat` config file to disable `--quit-if-one-screen`.

Does this happen for all terminal sizes?

**Comment by brainplot** [comment]
Sadly that doesn't seem the cause of the issue:

![image](https://user-images.githubusercontent.com/28014843/88466247-a4186800-ceca-11ea-81d1-3e3d6e839108.png)

**Comment by sharkdp** [comment]
Thank you for the feedback.

Could you please try to output `bat`s output to a file (`bat --color=always .\test.txt > output.txt`) and display it with `less` directly (or `less -R`)? Does this lead to the same problem?

**Comment by brainplot** [comment]
I tried that but it doesn't quite show the intended output:

![image](https://user-images.githubusercontent.com/28014843/88467184-b26b8180-ced4-11ea-9205-1a16da885c59.png)

However, even though the output is jumbled, it looks as though the bottom border was preserved (or at least it does so to me 😄)

**Comment by sharkdp** [comment]
It looks like `less` 562 fixes a lot of bugs on Windows: http://www.greenwoodsoftware.com/less/news.562.html

Some of them sound like they could be related to this issue (make screen repaint when "squished", Don't count lines in initial screen if using -X with -F), but that's just a wild guess.

A build of 562 by @jftuga is available here: https://github.com/jftuga/less-Windows/ 
It looks like it's not available via scoop so far.

**Comment by brainplot** [comment]
I believe you're right. I downloaded jftuga's build of `less` and used it and it looks like the bug is no longer there:

![image](https://user-images.githubusercontent.com/28014843/88489145-d8eff200-cf92-11ea-8e82-b6ee59634f76.png)

I guess we can close this. Thank you for your help!

**Comment by sharkdp** [comment]
Awesome, thank you very much for testing!

**Comment by brainplot** [comment]
I'm so sorry but it turns out we might need to reopen this 😄

Looks like setting `BAT_PAGER` is not a solution because **despite** setting it to `less -RF`, paging was disabled for some reason. As stated originally, paging is what breaks the output. Even with the updated version of `less` - which I can verify with `less --version` - the issue isn't solved.

**Comment by sharkdp** [comment]
Could you try to call `less --version` through the `--pager` option, to verify that `bat` is using the correct version?
```
bat --pager="less --version" …
```

**Comment by brainplot** [comment]
Looks like `bat` isn't honoring that...

![image](https://user-images.githubusercontent.com/28014843/89124454-5d80ca00-d4d7-11ea-8622-a1c6d06f0f2b.png)

The third command is the same as the second command but I hardcoded the path just to be extra sure.

**Comment by sharkdp** [comment]
Hm, that's weird. Seems like `bat` falls back to not paging as it cannot find the executable for some reason.

Is `less` in the PATH? (it should not be required, just curious)

What happens if you run
```
bat --pager="less --version" .\text.txt
```
or
```
bat --pager="less.exe --version" .\text.txt
```

Does it work if you run it from the directory in which the `less.exe` executable is placed?

**Comment by brainplot** [comment]
Ok so, the reason why the command interpolation didn't work is that `bat` doesn't understand Windows' path separator.
This command works as intended:
```powershell
$lessCommand = "$HOME\scoop\apps\less\current\less.exe" -replace '\\','/'
bat --pager="$lessCommand --version" $HOME\Projects\test.txt
```
It outputs:
```
less 562 (Spencer V8 regular expressions)
Copyright (C) 1984-2020  Mark Nudelman

less comes with NO WARRANTY, to the extent permitted by law.
For information about the terms of redistribution,
see the file named README in the less distribution.
Home page: http://www.greenwoodsoftware.com/less
```
Also, this works too as `less` is indeed in my `$PATH`
```powershell
bat --pager="less --version" $HOME\Projects\test.txt
```
with the same output as above!

**Comment by sharkdp** [comment]
ok.. thanks. but the original bug is not fixed, I assume.

Does it work if you manually pass `--no-init`/`-X` as well?
```
--pager="less -FRX"
```

**Comment by brainplot** [comment]
Nope, issue still remains!

**Comment by sharkdp** [comment]
This is a tough one. Maybe @jftuga or someone else using Windows could help out here?

Haven't heard this being reported by anyone else, so far.

**Comment by brainplot** [comment]
Yeah, let me know if you need any further testing on my part!

**Comment by sharkdp** [comment]
Could you please try out the latest master of `bat` ([bat-8e4b0b43-x86_64-pc-windows-gnu.zip.zip](https://github.com/sharkdp/bat/files/5791724/bat-8e4b0b43-x86_64-pc-windows-gnu.zip.zip)) and then post the output of
```
bat .\test.txt --diagnostic
```
or simply
```
bat --diagnostic
```
?

**Comment by eth-p** [comment]
I was able to still reproduce it on my Windows machine, so I can help out with the testing of this:

![image](https://user-images.githubusercontent.com/32112321/104109080-0c8fb180-527f-11eb-8723-b8234ee1cc1f.png)

#### Software version

bat 0.17.1

#### Operating system

Windows 6.2.9200

#### Command-line

```bash
".\\bat" --pager=../less test.rs --diagnostic
```

#### Environment variables

```bash
SHELL=<not set>
PAGER=<not set>
BAT_CACHE_PATH=<not set>
BAT_CONFIG_PATH=<not set>
BAT_OPTS=<not set>
BAT_PAGER=<not set>
BAT_STYLE=<not set>
BAT_TABS=<not set>
BAT_THEME=<not set>
XDG_CONFIG_HOME=<not set>
XDG_CACHE_HOME=<not set>
COLORTERM=<not set>
NO_COLOR=<not set>
```

#### Config file

Could not read contents of 'C:\Users\ethan\AppData\Roaming\bat\config': The system cannot find the path specified. (os error 3).

#### Compile time information

- Profile: release
- Target triple: x86_64-pc-windows-gnu
- Family: windows
- OS: windows
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-pc-windows-msvc

#### Less version

Could not run command 'less': The system cannot find the file specified. (os error 2)

Since less isn't installed globally for my example, here's the output of `.\less --version`:

> less 563 (Spencer V8 regular expressions)
> Copyright (C) 1984-2020  Mark Nudelman
> 
> less comes with NO WARRANTY, to the extent permitted by law.
> For information about the terms of redistribution,
> see the file named README in the less distribution.
> Home page: http://www.greenwoodsoftware.com/less

**Comment by sharkdp** [comment]
> I was able to still reproduce it on my Windows machine, so I can help out with the testing of this:

Great!

A few ideas:

- Just to make sure: everything works fine without a pager, right?
- What happens if you pipe the output of `bat --paging=always …` to a file? is the grid line missing as well?
- What happens if the `bat --decorations=always …` output is piped into `less`?

**Comment by eth-p** [comment]
My findings aren't entirely reproducible, unfortunately. It only cuts off the border *sometimes*, and I'm not exactly sure what's causing it. But interestingly enough though, I can consistently reproduce the issue when I explicitly use `--pager=less -RF`:
![image](https://user-images.githubusercontent.com/32112321/104109544-d6086580-5283-11eb-88f9-80fa610a17b5.png)

`less -R` works fine, so it seems the determining factor here is `--quit-if-one-screen`/`-F`. I can't figure out how to capture the output of `less` on Windows, so I'm not sure if `less -F` is just overwriting the final line, or if it's never printing it to begin with.

---

As for your questions:

> Just to make sure: everything works fine without a pager, right?

Correct.

> What happens if you pipe the output of bat --paging=always … to a file? is the grid line missing as well?

The output is identical between `--paging=always` and `--paging=never`.
These results are consistent with my inability to actually capture the output of `less` :(

> What happens if the bat --decorations=always … output is piped into less?

I don't think cmd.exe likes UTF-8.

![image](https://user-images.githubusercontent.com/32112321/104109505-714d0b00-5283-11eb-8532-8cb930994eff.png)

**Comment by sharkdp** [comment]
I tried to reproduce this as well and found some VERY weird behavior. Depending on whether I run:
```
bat test.txt
bat --pager="less -RF" test.txt
bat --pager=".\less -RF" test.txt
bat --pager=".\less.exe -RF" test.txt
```
`bat` chooses different versions of `less`.. and possibly different from the one that is reported by `bat --diagnostic`.

I then uninstalled the global "less" version and ran `bat` with
```
bat --pager=".\less.exe -RF" test.txt
```
and everything seems to work fine (with less 563). Weird things happened with less 551.


Could both of you please make sure that bat is actually calling less 563? Always make sure by running
```
bat --pager="path\to\less.exe --version" …
```

We should definitely make sure that we call `less` in the same way when performing the version check.

**Comment by eth-p** [comment]
I don't actually have a global `less` install on the machine I used to test it. When I used `less`, I did so by passing the relative path to the downloaded `less.exe` in the parent directory.

Also, that reminds me:

`bat --diagnostic --pager="..\less` couldn't  actually find `less`, and I had to paste the version manually. I haven't looked at the code for it yet, but I assume it's invoking the diagnostic printing before it's handling the --pager argument?

**Comment by sharkdp** [comment]
> I don't actually have a global `less` install on the machine I used to test it. When I used `less`, I did so by passing the relative path to the downloaded `less.exe` in the parent directory.

ok. Could still be useful to make sure it's properly called using `bat --pager="..\less --version"`. I actually had to use the `.exe` extension. Does it work without `.exe` for you?

**Comment by sharkdp** [comment]
> `bat --diagnostic --pager="..\less` couldn't actually find `less`, and I had to paste the version manually. I haven't looked at the code for it yet, but I assume it's invoking the diagnostic printing before it's handling the --pager argument?

Yes :-/. It simply runs`less --version` without considering the value of `--pager`/`PAGER`/`BAT_PAGER`. Similar to what the diagnostics script did so far.

**Comment by sharkdp** [comment]
What's more problematic is this in `less.rs`:
``` rust
pub fn retrieve_less_version() -> Option<usize> {
    let cmd = Command::new("less").arg("--version").output().ok()?;
    parse_less_version(&cmd.stdout)
}
```

If you pass in a custom `less` binary via `--pager`, the version check will not be affected. This means that it will probably fail (return `None`), which means that we add the `--no-init` argument which is not needed for newer less versions.

**Comment by eth-p** [comment]
Looking at the screenshots above, it turns out my above commands actually used `../less` and not `..\less`.
That seems to be a happy accident, since it turns out using `..\less[.exe]` doesn't actually work 😅 

```
X:\Downloads\win-pager-test>.\bat --pager="../less --version" --paging=always -A output-no-pager.txt
less 563 (Spencer V8 regular expressions)
Copyright (C) 1984-2020  Mark Nudelman

less comes with NO WARRANTY, to the extent permitted by law.
For information about the terms of redistribution,
see the file named README in the less distribution.
Home page: http://www.greenwoodsoftware.com/less

X:\Downloads\win-pager-test>.\bat --pager="../less.exe --version" --paging=always -A output-no-pager.txt
less 563 (Spencer V8 regular expressions)
Copyright (C) 1984-2020  Mark Nudelman

less comes with NO WARRANTY, to the extent permitted by law.
For information about the terms of redistribution,
see the file named README in the less distribution.
Home page: http://www.greenwoodsoftware.com/less

X:\Downloads\win-pager-test>.\bat --pager="..\less.exe --version" --paging=always -A output-no-pager.txt
───────┬────────────────────────────────────────────────────────────────────────────────────────
       │ File: output-no-pager.txt
───────┼────────────────────────────────────────────────────────────────────────────────────────
   1   │ <manually truncated for space>
───────┴────────────────────────────────────────────────────────────────────────────────────────

X:\Downloads\win-pager-test>.\bat --pager="..\\less.exe --version" --paging=always -A output-no-pager.txt
less 563 (Spencer V8 regular expressions)
Copyright (C) 1984-2020  Mark Nudelman

less comes with NO WARRANTY, to the extent permitted by law.
For information about the terms of redistribution,
see the file named README in the less distribution.
Home page: http://www.greenwoodsoftware.com/less

X:\Downloads\win-pager-test>
```

---

> Yes :-/. It simply runsless --version without considering the value of --pager/PAGER/BAT_PAGER. Similar to what the diagnostics script did so far.

Seems like a good opportunity for us to improve on something the original shell script couldn't reasonably do :)

> What's more problematic is this in less.rs:
> 
> ```rs
> pub fn retrieve_less_version() -> Option<usize> {
>     let cmd = Command::new("less").arg("--version").output().ok()?;
>     parse_less_version(&cmd.stdout)
> }
> ```

That *is* problematic.

**Comment by sharkdp** [comment]
@eth-p: After merging #1509, could you maybe have another look at this and summarize the current status?

**Comment by brainplot** [comment]
Since I was following this PR, I just tried to build from `HEAD` (i.e. commit `9ad401b`) and the problem is still there:

<p  align="center"><img width="580" alt="bat" src="https://user-images.githubusercontent.com/28014843/108039273-2cfc1a00-703c-11eb-8507-97a1c81c35a0.png"></p>

I don't know if I said this before but it happens only with short files that don't need to be paged. I tried to run the repo's `Cargo.toml` through `bat` and I get the file paged and I can scroll all the way to the bottom and I can see the bottom border just fine. Again, I don't remember if this was brought up in the original issue.

**Comment by sharkdp** [comment]
@brainplot Could you please execute `.\target\release\bat.exe .\.gitignore --diagnostic` and post the output here?

**Comment by brainplot** [comment]
#### Software version

bat 0.17.1

#### Operating system

Windows 6.2.9200

#### Command-line

```bash
"C:\\Users\\Gianluca\\Projects\\bat\\target\\release\\bat.exe" ".\\.gitignore" --diagnostic
```

#### Environment variables

```bash
SHELL=<not set>
PAGER=<not set>
BAT_PAGER=<not set>
BAT_CACHE_PATH=<not set>
BAT_CONFIG_PATH=<not set>
BAT_OPTS=<not set>
BAT_STYLE=<not set>
BAT_TABS=<not set>
BAT_THEME=<not set>
XDG_CONFIG_HOME=<not set>
XDG_CACHE_HOME=<not set>
COLORTERM=<not set>
NO_COLOR=<not set>
MANPAGER=<not set>
```

#### Config file

```
#!/hint/sh

# Set the theme
--theme=TwoDark

# Use italic text on the terminal (not supported on all terminals)
--italic-text=always

# Use C++ syntax (instead of C) for .h header files
--map-syntax h:C++

# Open .clang-format files as YAML files
--map-syntax .clang-format:YAML

# Open cargo config files as TOML files
--map-syntax **/*cargo/config:TOML

# Open tmux config file with bash syntax
--map-syntax .tmux.conf:'Bourne Again Shell (bash)'

# Open ebuild files used in gentoo with bash syntax
--map-syntax *.ebuild:'Bourne Again Shell (bash)'
```

#### Compile time information

- Profile: release
- Target triple: x86_64-pc-windows-msvc
- Family: windows
- OS: windows
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-pc-windows-msvc

#### Less version

```
> less --version
less 563 (Spencer V8 regular expressions)
Copyright (C) 1984-2020  Mark Nudelman

less comes with NO WARRANTY, to the extent permitted by law.
For information about the terms of redistribution,
see the file named README in the less distribution.
Home page: http://www.greenwoodsoftware.com/less
```

**Comment by sharkdp** [comment]
Long shot: Could you try with `bat --no-config`? Or manually comment-out the `--italic-text=always` setting? Maybe that causes problems.

I just now realize that you have a multiline prompt. Could you try to temporarily set that to a single-line prompt and see if it resolves the problem? Maybe your shell is overwriting that last line of `bat`s output somehow.

**Comment by jftuga** [comment]
@brainplot 

I just created a `pre-release` of `less-579` with Windows binaries.

https://github.com/jftuga/less-Windows/releases/tag/less-v579.0

Please try out this version and let us know if this fixes your problem.

**Comment by brainplot** [comment]
Hi @jftuga, the version of `less` you suggested indeed fixes the issue.

![image](https://user-images.githubusercontent.com/28014843/108243529-7ccb0600-714e-11eb-91bf-fb04b968a3d7.png)

To make sure it wasn't a temporary setting I had changed for `bat`, I tried to delete your version of `less` from the temporary folder I had saved it in and the issue with the bottom border was occurring again. Thanks for looking into this!
